File: ethcore/src/trace.rs
// Copyright 2015, 2016 Ethcore (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

//! Tracing datatypes.
use common::*;

/// Description of a _call_ action, either a `CALL` operation or a message transction.
#[derive(Debug, Clone, PartialEq)]
pub struct TraceCall {
	/// The sending account.
	pub from: Address,
	/// The destination account.
	pub to: Address,
	/// The value transferred to the destination account.
	pub value: U256,
	/// The gas available for executing the call.
	pub gas: U256,
	/// The input data provided to the call.
	pub input: Bytes,
	/// The result of the operation; the gas used and the output data of the call.
	pub result: Option<(U256, Bytes)>,
}

/// Description of a _create_ action, either a `CREATE` operation or a create transction.
#[derive(Debug, Clone, PartialEq)]
pub struct TraceCreate {
	/// The address of the creator.
	pub from: Address,
	/// The value with which the new account is endowed.
	pub value: U256,
	/// The gas available for the creation init code.
	pub gas: U256,
	/// The init code.
	pub init: Bytes,
	/// The result of the operation; tuple of the gas used, the address of the newly created account and its code.
	/// NOTE: Presently failed operations are not reported so this will always be `Some`.
	pub result: Option<(U256, Address, Bytes)>,
//	pub output: Bytes,
}

/// Description of an action that we trace; will be either a call or a create.
#[derive(Debug, Clone, PartialEq)]
pub enum TraceAction {
	/// Action isn't yet known.
	Unknown,
	/// It's a call action.
	Call(TraceCall),
	/// It's a create action.
	Create(TraceCreate),
}

#[derive(Debug, Clone, PartialEq)]
/// A trace; includes a description of the action being traced and sub traces of each interior action.
pub struct Trace {
	/// The number of EVM execution environments active when this action happened; 0 if it's
	/// the outer action of the transaction.
	pub depth: usize,
	/// The action being performed.
	pub action: TraceAction,
	/// The sub traces for each interior action performed as part of this call.
	pub subs: Vec<Trace>,
}

impl Default for Trace {
	fn default() -> Trace {
		Trace {
			depth: 0,
			action: TraceAction::Unknown,
			subs: vec![],
		}
	}
}

impl TraceAction {
	/// Compose a `TraceAction` from an `ActionParams`, knowing that the action is a call.
	pub fn from_call(p: &ActionParams) -> TraceAction {
		TraceAction::Call(TraceCall {
			from: p.sender.clone(),
			to: p.address.clone(),
			value: match p.value { ActionValue::Transfer(ref x) | ActionValue::Apparent(ref x) => x.clone() },
			gas: p.gas.clone(),
			input: p.data.clone().unwrap_or(vec![]),
			result: None,
		})
	}

	/// Compose a `TraceAction` from an `ActionParams`, knowing that the action is a create.
	pub fn from_create(p: &ActionParams) -> TraceAction {
		TraceAction::Create(TraceCreate {
			from: p.sender.clone(),
			value: match p.value { ActionValue::Transfer(ref x) | ActionValue::Apparent(ref x) => x.clone() },
			gas: p.gas.clone(),
			init: p.code.clone().unwrap_or(vec![]),
			result: None,
		})
	}
}

File: js/src/jsonrpc/interfaces/trace.js
// Copyright 2015, 2016 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import { BlockNumber, Data, Hash, Integer } from '../types';
import { withPreamble, Dummy } from '../helpers';

const SECTION_FILTERING = 'Transaction-Trace Filtering';
const SECTION_ADHOC = 'Ad-hoc Tracing';

export default withPreamble(`

The trace module is for getting a deeper insight into transaction processing.
It includes two sets of calls; the transaction trace filtering API and the ad-hoc tracing API.

## The Ad-hoc Tracing API

The ad-hoc tracing API allows you to perform a number of different diagnostics on calls or transactions,
eitherhistorical ones from the chain or hypothetical ones not yet mined. The diagnostics include:

- \`trace\` **Transaction trace**. An equivalent trace to that in the previous section.
- \`vmTrace\` **Virtual Machine execution trace**. Provides a full trace of the VM's state throughout the execution of the transaction, including for any subcalls.
- \`stateDiff\` **State difference**. Provides information detailing all altered portions of the Ethereum state made due to the execution of the transaction.

There are three means of providing a transaction to execute; either providing the same information as when making
a call using \`eth_call\` (see \`trace_call\`), through providing raw, signed, transaction data as when using
\`eth_sendRawTransaction\` (see \`trace_rawTransaction\`) or simply a transaction hash for a previously mined
transaction (see \`trace_replayTransaction\`). In the latter case, your node must be in archive mode or the
transaction should be within the most recent 1000 blocks.

## The Transaction-Trace Filtering API

These APIs allow you to get a full *externality* trace on any transaction executed throughout the Parity chain.
Unlike the log filtering API, you are able to search and filter based only upon address information.
Information returned includes the execution of all \`CREATE\`s, \`SUICIDE\`s and all variants of \`CALL\` together
with input data, output data, gas usage, amount transferred and the success status of each individual action.

In order to use these API Parity must be fully synced with flags \`$ parity --tracing on\`.

### \`traceAddress\` field

The \`traceAddress\` field of all returned traces, gives the exact location in the call trace [index in root,
index in first \`CALL\`, index in second \`CALL\`, ...].

i.e. if the trace is:
\`\`\`
A
  CALLs B
    CALLs G
  CALLs C
    CALLs G
\`\`\`
then it should look something like:

\`[ {A: []}, {B: [0]}, {G: [0, 0]}, {C: [1]}, {G: [1, 0]} ]\`

`, {
  block: {
    section: SECTION_FILTERING,
    desc: 'Returns traces created at given block',
    params: [
      {
        type: BlockNumber,
        desc: 'Integer block number, or \'latest\' for the last mined block or \'pending\', \'earliest\' for not yet mined transactions'
      }
    ],
    returns: {
      type: Array,
      desc: 'Block traces'
    }
  },

  filter: {
    section: SECTION_FILTERING,
    desc: 'Returns traces matching given filter',
    params: [
      {
        type: Object,
        desc: 'The filter object'
      }
    ],
    returns: {
      type: Array,
      desc: 'Traces matching given filter'
    }
  },

  get: {
    section: SECTION_FILTERING,
    desc: 'Returns trace at given position.',
    params: [
      {
        type: Hash,
        desc: 'Transaction hash'
      },
      {
        type: Integer,
        desc: 'Trace position witing transaction'
      }
    ],
    returns: {
      type: Object,
      desc: 'Trace object'
    }
  },

  transaction: {
    section: SECTION_FILTERING,
    desc: 'Returns all traces of given transaction',
    params: [
      {
        type: Hash,
        desc: 'Transaction hash'
      }
    ],
    returns: {
      type: Array,
      desc: 'Traces of given transaction',
      example: new Dummy('[ ... ]')
    }
  },

  call: {
    section: SECTION_ADHOC,
    desc: 'Returns traces for a specific call',
    params: [
      {
        type: Object,
        desc: 'Call options'
      },
      {
        type: BlockNumber,
        desc: 'The blockNumber'
      },
      {
        type: Array,
        desc: 'Type of trace, one or more of \'vmTrace\', \'trace\' and/or \'stateDiff\''
      }
    ],
    returns: {
      type: Array,
      desc: 'Block traces'
    }
  },

  rawTransaction: {
    section: SECTION_ADHOC,
    desc: 'Traces a call to eth_sendRawTransaction without making the call, returning the traces',
    params: [
      {
        type: Data,
        desc: 'Transaction data'
      },
      {
        type: Array,
        desc: 'Type of trace, one or more of \'vmTrace\', \'trace\' and/or \'stateDiff\''
      }
    ],
    returns: {
      type: Array,
      desc: 'Block traces'
    }
  },

  replayTransaction: {
    section: SECTION_ADHOC,
    desc: 'Replays a transaction, returning the traces',
    params: [
      {
        type: Hash,
        desc: 'Transaction hash'
      },
      {
        type: Array,
        desc: 'Type of trace, one or more of \'vmTrace\', \'trace\' and/or \'stateDiff\''
      }
    ],
    returns: {
      type: Array,
      desc: 'Block traces'
    }
  }
});
